### PR TITLE
fix: prevent damaged token embeddings from dominating grad clipping

### LINKED ISSUE #3174: Numerically robust clip norm and rare tokens repair

### PR DESCRIPTION
**Pull request by akoumpa**
## Summary

Prevent rare damaged input-embedding rows from overwhelming global gradient clipping during CPT, while retaining overflow-safe gradient-norm reporting.

## Root cause

The Nemotron-3 Nano 4B checkpoint has 79 input-embedding rows with L2 norm at or below `1e-4` (73 are exactly zero). When full-vocabulary mock data reaches these tokens, their embedding gradients can reach `6.21e31`.

An overflow-safe norm correctly reports the resulting global norm near `3e32`, but ordinary global clipping to `0.3` then multiplies every model gradient by roughly `1e-33`. That avoids an `inf` value but effectively discards the useful update.

## Changes

- Compute global gradient norms with max scaling and FP64 accumulation, including DTensor-aware reduction.
- Add opt-in `embedding_row_repair` configuration applied once after checkpoint loading and before optimizer construction.
- Identify non-finite or near-zero input-embedding rows and log their global token IDs.
- Replace each damaged row with the corresponding healthy `lm_head` direction, scaled to the RMS norm of healthy input rows.
- Handle replicated, vocabulary-sharded, and hidden-dimension-sharded DTensors without gathering the full embedding table.
- Fail before mutation when the number of damaged rows exceeds `max_rows`, the source output row is also damaged, or layouts are incompatible.
- Enable the repair for the Nemotron Nano 4B example with `min_norm: 1.0e-4` and document the option.

The repair is a one-time startup operation. It does not scan or alter embeddings on every training step.

## Validation

Local:

- `pytest -q tests/unit_tests/components/training/test_embedding_row_repair.py tests/unit_tests/training/test_train_utils.py tests/unit_tests/config/test_example_yaml_linter.py`: 44 passed
- `pytest -q tests/unit_tests/components/training`: 69 passed
- Targeted Ruff formatting and lint checks passed.

cw-dfw, 1 node / 8 H100s, FSDP2, full-vocabulary mock CPT:

| | Step 0 | Step 1 |
|---|---:|---:|
| Before repair, grad norm | `3.0889764e32` | `2.0009568e32` |
| After repair, grad norm | `21.1992` | `15.5722` |
| After repair, loss | `13.7071` | `12.9191` |

The run identified and repaired 79 rows with a healthy-row target norm of `0.77678`. Both steps completed with finite losses and gradients.

The broader `test_train_ft.py` module could not collect in the local system environment because its installed MLflow version expects a different `cachetools` API. The actual recipe setup ordering and multi-rank DTensor path were exercised by the cw-dfw run.

### CODE REVIEW

File: nemo_automodel/components/training/embedding_row_repair.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+        placements=placements,
+    )
+
+
+def _full_row_squared_norms(view: _WeightView) -> torch.Tensor:
+    local = view.local.detach().to(torch.float32)
+    row_squared_norms = local.square().sum(dim=1, dtype=torch.float64)
+    if view.mesh is None:
+        return row_squared_norms
+
```
[the remark this answers is not in the document]
Comment (athitten) [reply nemo_automodel/components/training/embedding_row_repair.py:170]: @akoumpa this might be good to add to maintain code quality

File: tests/unit_tests/components/training/test_embedding_row_repair.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    recipe = RecipeConfig(ConfigNode({"embedding_row_repair": {"min_norm": 1.0e-4, "max_rows": 8}}))
+    assert recipe.embedding_row_repair == EmbeddingRowRepairConfig(min_norm=1.0e-4, max_rows=8)
+    assert EmbeddingRowRepairConfig(enabled=False).apply(_TinyLM()) is None
+
+
+def test_repair_supports_single_rank_sharded_dtensors(damaged_model, single_rank_gloo):
+    from torch.distributed.device_mesh import init_device_mesh
+    from torch.distributed.tensor import Shard, distribute_tensor
+
+    mesh = init_device_mesh("cpu", (1,))
```
[the remark this answers is not in the document]
Comment (athitten) [reply tests/unit_tests/components/training/test_embedding_row_repair.py:164]: @akoumpa can we add a 2 GPU test ?

**Comment by athitten** [comment]
@akoumpa can we also verify loss, grad_norms continue to remain sane for larger steps like 100/200 ? Thanks

**Comment by akoumpa** [comment]
Verified the longer-step sanity run on cw-dfw against head `3c6241e89608e516a128b4af29d5c7eb60a6d9fb`.

Config summary: 1 node / 8 H100s, FSDP2, full-vocabulary mock data (vocab_size=131072), seq_len=128, local/global batch 1/8, max_steps=200. I set `model.use_mamba_kernels=false` for this validation run to avoid the remote Triton/Mamba first-step autotune/cache path; the embedding repair and grad-norm code paths are unchanged.

Repair log: repaired 79 input embedding rows, minimum_before=0, target_norm=7.768e-01.

| step | loss | grad_norm |
|---:|---:|---:|
| 0 | 13.6376 | 22.2936 |
| 100 | 11.9176 | 2.9912 |
| 199 | 11.9158 | 2.8966 |

Parsed 200 metric rows from the run log; all loss and grad_norm values were finite. grad_norm range: 2.7562 to 27.6283. Final checkpoint export completed successfully.

**Review by athitten — APPROVED** [review]
LGTM, thank you @akoumpa !
